File: ai_remixmate_feature_lab/typescript/src/neural/featureRegistry.ts
import type { NeuralFeatureSpec } from "./NeuralTypes";

export const TRACK_VECTOR_SIZE = 16;
export const PAIR_VECTOR_SIZE = 40;

export const NEURAL_FEATURE_SPECS: Record<string, NeuralFeatureSpec> = {
  bpm_compatibility: {
    name: "bpm_compatibility",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns tempo blend quality from feedback."
  },
  key_compatibility: {
    name: "key_compatibility",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns harmonic acceptability beyond simplified Camelot rules."
  },
  energy_compatibility: {
    name: "energy_compatibility",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns preferred energy-flow behavior."
  },
  timbre_compatibility: {
    name: "timbre_compatibility",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns perceived sonic texture fit."
  },
  vocal_clash_risk: {
    name: "vocal_clash_risk",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns vocal overlap risk from render reviews."
  },
  compatibility_score: {
    name: "compatibility_score",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 16,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns overall transition approval."
  },
  transition_planning: {
    name: "transition_planning",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 16,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns whether generated cue timing was accepted."
  },
  remix_recipe_quality: {
    name: "remix_recipe_quality",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 16,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns generated recipe usefulness."
  },
  automix_next_track: {
    name: "automix_next_track",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 16,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns set-order approval."
  },
  track_match: {
    name: "track_match",
    inputSize: PAIR_VECTOR_SIZE,
    hiddenSize: 16,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns recommendation relevance."
  },
  beatgrid_confidence: {
    name: "beatgrid_confidence",
    inputSize: TRACK_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns whether beatgrids need correction."
  },
  stem_quality: {
    name: "stem_quality",
    inputSize: TRACK_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns stem usefulness and bleed risk."
  },
  waveform_interest: {
    name: "waveform_interest",
    inputSize: TRACK_VECTOR_SIZE,
    hiddenSize: 12,
    outputSize: 1,
    learningRate: 0.05,
    version: 1,
    description: "Learns section salience for cues and loops."
  }
};

export function neuralFeatureNames(): string[] {
  return Object.keys(NEURAL_FEATURE_SPECS).sort();
}

File: ai_remixmate_feature_lab/python/ai_remixmate_features/neural/feature_registry.py
from __future__ import annotations

from .core import NeuralFeatureSpec, TinyOnlineMLP
from .vectorizers import PAIR_VECTOR_SIZE, TRACK_VECTOR_SIZE


FEATURE_MODEL_SPECS: dict[str, NeuralFeatureSpec] = {
    "bpm_compatibility": NeuralFeatureSpec(
        name="bpm_compatibility",
        input_size=PAIR_VECTOR_SIZE,
        description="Learns tempo blend quality from user feedback and manual corrections.",
    ),
    "key_compatibility": NeuralFeatureSpec(
        name="key_compatibility",
        input_size=PAIR_VECTOR_SIZE,
        description="Learns harmonic acceptability beyond simplified Camelot rules.",
    ),
    "energy_compatibility": NeuralFeatureSpec(
        name="energy_compatibility",
        input_size=PAIR_VECTOR_SIZE,
        description="Learns user preference for energy-flow similarity or contrast.",
    ),
    "timbre_compatibility": NeuralFeatureSpec(
        name="timbre_compatibility",
        input_size=PAIR_VECTOR_SIZE,
        description="Learns perceived sonic texture fit.",
    ),
    "vocal_clash_risk": NeuralFeatureSpec(
        name="vocal_clash_risk",
        input_size=PAIR_VECTOR_SIZE,
        description="Learns vocal conflict risk from render reviews.",
    ),
    "compatibility_score": NeuralFeatureSpec(
        name="compatibility_score",
        input_size=PAIR_VECTOR_SIZE,
        hidden_size=16,
        description="Learns overall transition approval as a neural adjustment layer.",
    ),
    "transition_planning": NeuralFeatureSpec(
        name="transition_planning",
        input_size=PAIR_VECTOR_SIZE,
        hidden_size=16,
        description="Learns whether generated entry and exit timing was accepted.",
    ),
    "remix_recipe_quality": NeuralFeatureSpec(
        name="remix_recipe_quality",
        input_size=PAIR_VECTOR_SIZE,
        hidden_size=16,
        description="Learns whether generated beginner-readable recipe steps were useful.",
    ),
    "automix_next_track": NeuralFeatureSpec(
        name="automix_next_track",
        input_size=PAIR_VECTOR_SIZE,
        hidden_size=16,
        description="Learns next-track ordering confidence from set-builder feedback.",
    ),
    "track_match": NeuralFeatureSpec(
        name="track_match",
        input_size=PAIR_VECTOR_SIZE,
        hidden_size=16,
        description="Learns recommendation relevance from click, skip, and queue actions.",
    ),
    "beatgrid_confidence": NeuralFeatureSpec(
        name="beatgrid_confidence",
        input_size=TRACK_VECTOR_SIZE,
        description="Learns whether an analyzed beatgrid needs manual correction.",
    ),
    "stem_quality": NeuralFeatureSpec(
        name="stem_quality",
        input_size=TRACK_VECTOR_SIZE,
        description="Learns whether available stems are clean enough for stem-aware remixing.",
    ),
    "waveform_interest": NeuralFeatureSpec(
        name="waveform_interest",
        input_size=TRACK_VECTOR_SIZE,
        description="Learns which track regions tend to become cues, loops, or transition points.",
    ),
}


def feature_names() -> list[str]:
    return sorted(FEATURE_MODEL_SPECS)


def create_model_registry() -> dict[str, TinyOnlineMLP]:
    return {name: TinyOnlineMLP(spec) for name, spec in FEATURE_MODEL_SPECS.items()}
